build.rs

```diff
@@ -42,36 +42,147 @@ fn main() {
         .expect("OUT_DIR should be set by cargo")
         .into();
 
-    let template_dir = manifest_dir.join("templates").join("new");
-    let dst_file = out_dir.join("template.zip");
+    generate_cargo_keys();
+    rerun_if_git_head_changed();
 
+    let template_dir = manifest_dir.join("templates").join("new");
+    let template_dst_file = out_dir.join("template.zip");
     println!(
         "Creating template zip: template_dir '{}', destination archive '{}'",
         template_dir.display(),
-        dst_file.display()
+        template_dst_file.display()
     );
-
-    generate_cargo_keys();
-    rerun_if_git_head_changed();
-
-    std::process::exit(
-        match zip_dir(&template_dir, &dst_file, CompressionMethod::Stored) {
+    let zipped_template =
+        match zip_dir(&template_dir, &template_dst_file, CompressionMethod::Stored) {
             Ok(_) => {
                 println!(
                     "done: {} written to {}",
                     template_dir.display(),
-                    dst_file.display()
+                    template_dst_file.display()
                 );
-                0
+                true
             }
             Err(e) => {
                 eprintln!("Error: {:?}", e);
-                1
+                false
             }
-        },
+        };
+
+    // This zip file will contain the `dylint` driver, this is one file named in the form of
+    // `libink_linting@nightly-2021-11-04-x86_64-unknown-linux-gnu.so`. This file is obtained
+    // by building the crate in `ink_linting/`.
+    let dylint_driver_dst_file = out_dir.join("ink-dylint-driver.zip");
+
+    #[cfg(feature = "cargo-clippy")]
+    {
+        // For `clippy` runs it is not necessary to build the `dylint` driver.
+        // Furthermore the fixed Rust nightly specified in `ink_linting/rust-toolchain`
+        // contains a bug that results in an `error[E0786]: found invalid metadata files` ICE.
+        //
+        // We still have to create an empty file though, due to the `include_bytes!` macro.
+        File::create(dylint_driver_dst_file).unwrap_or_else(|err| {
+            panic!(
+                "Failed creating an empty ink-dylint-driver.zip file: {:?}",
+                err
+            );
+        });
+        std::process::exit(bool_to_exit_code(zipped_template));
+    }
+
+    #[cfg(not(feature = "cargo-clippy"))]
+    {
+        let zipped_lints = build_dylint_driver(manifest_dir, out_dir, dylint_driver_dst_file)
+            .map(|_| true)
+            .unwrap_or_else(|err| {
+                panic!("Failed building dylint driver: {:?}", err);
+            });
+        std::process::exit(bool_to_exit_code(zipped_template && zipped_lints));
```

Comment:
I would just replace this whole block with a call to `build_dylint_driver` and then add an alternative with `#[cfg(feature = "cargo-clippy")]` since you already have that method decorated. It will make the control flow easier to follow